snow/engine/snowman/bootstrap/storage.go

```diff
@@ -0,0 +1,252 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package bootstrap
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/consensus/snowman"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/snow/engine/snowman/block"
+	"github.com/ava-labs/avalanchego/snow/engine/snowman/bootstrap/interval"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/utils/timer"
+)
+
+const (
+	batchWritePeriod      = 64
+	iteratorReleasePeriod = 1024
+	logPeriod             = 5 * time.Second
+)
+
+// getMissingBlockIDs returns the ID of the blocks that should be fetched to
+// attempt to make a single continuous range from
+// (lastAcceptedHeight, highestTrackedHeight].
+//
+// For example, if the tree currently contains heights [1, 4, 6, 7] and the
+// lastAcceptedHeight is 2, this function will return the IDs corresponding to
+// blocks [3, 5].
+func getMissingBlockIDs(
+	ctx context.Context,
+	db database.KeyValueReader,
+	parser block.Parser,
+	tree *interval.Tree,
+	lastAcceptedHeight uint64,
+) (set.Set[ids.ID], error) {
+	var (
+		missingBlocks     set.Set[ids.ID]
+		intervals         = tree.Flatten()
+		lastHeightToFetch = lastAcceptedHeight + 1
+	)
+	for _, i := range intervals {
+		if i.LowerBound <= lastHeightToFetch {
+			continue
+		}
+
+		blkBytes, err := interval.GetBlock(db, i.LowerBound)
+		if err != nil {
+			return nil, err
+		}
+
+		blk, err := parser.ParseBlock(ctx, blkBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		parentID := blk.Parent()
+		missingBlocks.Add(parentID)
+	}
+	return missingBlocks, nil
+}
+
+// process a series of consecutive blocks starting at [blk].
+//
+//   - blk is a block that is assumed to have been marked as acceptable by the
+//     bootstrapping engine.
+//   - ancestors is a set of blocks that can be used to lookup blocks.
+//
+// If [blk]'s height is <= the last accepted height, then it will be removed
+// from the missingIDs set.
+//
+// Returns a newly discovered blockID that should be fetched.
+func process(
```

Comment:
q: IIUC we don't check whether blocks received in an `Ancestors` messages are contiguous in `block.BatchedParseBlock. It's not up to the VM to check block contiguity and we don't really care. Instead it's `process` responsability to find what's next to download (included possible gaps in the Ancestors message). Do I understand correctly?

Reply:
Also in case the ancestors blocks have multiple gaps, we stop at the first one right? We won't greedly record the other height and return multiple heights to fetch.

Reply:
One thing that `process` doesn't do - which it could do... is process blocks inside of `ancestors` that are contained inside of `missingBlockIDs`. I felt like this wasn't worth implementing because:

1. It introduces more complexity
2. This will only store blocks in a couple of niche cases

Cases this could be useful:

1. If a node provides non-consecutive blocks in ancestors but during the start of bootstrapping we have added one of the provided blocks as desired.
    a. This can only happen in the byzantine node case.
2. A node provides a set of consecutive ancestors where only some of the ancestors have already been persisted.
    a. We'll have already sent a request for this ancestors, so this is only an optimization if that request were to fail (as we could then decide not to re-issue the request).